README.md
![GitHub release (latest by date)](https://img.shields.io/github/v/release/ImageProcessing-ElectronicPublications/jpeg-recompress)
![GitHub Release Date](https://img.shields.io/github/release-date/ImageProcessing-ElectronicPublications/jpeg-recompress)
![GitHub repo size](https://img.shields.io/github/repo-size/ImageProcessing-ElectronicPublications/jpeg-recompress)
![GitHub all releases](https://img.shields.io/github/downloads/ImageProcessing-ElectronicPublications/jpeg-recompress/total)
![GitHub](https://img.shields.io/github/license/ImageProcessing-ElectronicPublications/jpeg-recompress)

# JPEG Recompress

Utilities for archiving photos for saving to long term storage or serving over the web. The goals are:

 * Use a common, well supported format (JPEG)
 * Minimize storage space and cost
 * Identify duplicates / similar photos

Approach:

 * Command line utilities and scripts
 * Simple options and useful help
 * Good quality output via sane defaults

Contributions to this project are very welcome.

## Download

You can download the latest source and binary releases from the [JPEG Recompress releases page](https://github.com/ImageProcessing-ElectronicPublications/jpeg-recompress/releases).

## Utilities

The following utilities are part of this project. All of them accept a `--help` parameter to see the available options.

### jpeg-recompress

Compress JPEGs by re-encoding to the smallest JPEG quality while keeping _perceived_ visual quality the same and by making sure huffman tables are optimized. This is a __lossy__ operation, but the images are visually identical and it usually saves 30-70% of the size for JPEGs coming from a digital camera, particularly DSLRs. By default all EXIF/IPTC/XMP and color profile metadata is copied over, but this can be disabled to save more space if desired.

There is no need for the input file to be a JPEG. In fact, you can use `jpeg-recompress` as a replacement for `cjpeg` by using PPM input and the `--ppm` option.

The better the quality of the input image is, the better the output will be.

Some basic photo-related editing options are available, such as removing fisheye lens distortion.

#### Demo

Below are two 100% crops of [Nikon's D3x Sample Image 2](http://static.nikonusa.com/D3X_gallery/index.html). The left shows the original image from the camera, while the others show the output of `jpeg-recompress` with the `medium` quality setting and various comparison methods. By default SSIM is used, which lowers the file size by **88%**. The recompression algorithm chooses a JPEG quality of 80. By comparison the `veryhigh` quality setting chooses a JPEG quality of 93 and saves 70% of the file size.

![JPEG recompression comparison](https://cloud.githubusercontent.com/assets/106826/3633843/5fde26b6-0eff-11e4-8c98-f18dbbf7b510.png)

Why are they different sizes? The default quality settings are set to average out to similar visual quality over large data sets. They may differ on individual photos (like above) because each metric considers different parts of the image to be more or less important for compression.

#### Image Comparison Metrics

The following metrics are available when using `jpeg-recompress`. SUM is the default.

Name        | Option        | Description
----------- | ------------- | -----------
MPE         | `-m mpe`      | Mean pixel error (as used by [imgmin](https://github.com/rflynn/imgmin))
PSNR        | `-m psnr`     | [Peak signal-to-noise ratio](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio)
MSE         | `-m mse`      | [Mean squared error](https://en.wikipedia.org/wiki/Mean_squared_error)
MSEF        | `-m msef`     | `sqrt(MSE/Variance)`
Correlation | `-m cor`      | [Correlation](https://en.wikipedia.org/wiki/Correlation)
SSIM        | `-m ssim`     | [Structural similarity](http://en.wikipedia.org/wiki/Structural_similarity)
MS-SSIM*    | `-m ms-ssim`  | Multi-scale structural similarity (slow!) ([2008 paper](http://foulard.ece.cornell.edu/publications/dmr_hvei2008_paper.pdf))
VIFP1       | `-m vifp1`    | [The visual information fidelity (VIF)](http://live.ece.utexas.edu/publications.php) 1 layer.
SmallFry    | `-m smallfry` | Linear-weighted BBCQ-like ([original project](https://github.com/dwbuiten/smallfry), [2011 BBCQ paper](http://spie.org/Publications/Proceedings/Paper/10.1117/12.872231) -> [LibSmallFry](https://github.com/ImageProcessing-ElectronicPublications/libsmallfry))
SharpenBad  | `-m shbad`    | Sharpen discrepancies ([LibSmallFry](https://github.com/ImageProcessing-ElectronicPublications/libsmallfry))
NHW         | `-m nhw`      | NHW convolutional metric ([original project](https://github.com/rcanut/NHW_Neatness_Metrics) -> [LibSmallFry](https://github.com/ImageProcessing-ElectronicPublications/libsmallfry))
1 pair      | `-m ssimfry`  | `(ssim + smallfry) / 2`
2 pair      | `-m ssimshb`  | `(ssim + shbad) / 2`
SUMMARY     | `-m sum`      | `(ssim + vipf1 + smallfry + shbad + nhw) / 5` **DEFAULT**

**Note**: The SmallFry algorithm may be [patented](http://www.jpegmini.com/main/technology) so use with caution.

#### "Universal Scale" of metrics (UM):

![UM example](plot_um.svg)

```
  0.0
  ... (DIRTY) ...
  0.5
  ... (LOW) ...
  0.75
  ... (MEDIUM) ...
  0.875
  ... (SUBHIGH) ...
  0.9375
  ... (HIGH) ...
  0.96875
  ... (VERYHIGH) ...
  1.0
```
Trends:
```
  UM = 0.29 * (sqrt(sqrt(255.0 / MPE)) - 1.0)
  UM = 0.557 * (sqrt(PNSR) - 5.0)
  UM = 0.5 * (sqrt(sqrt(1.0 / MSEF)) - 1.0)
  UM = 1.0 * cor_sigma(cor_sigma(COR))
  UM = 1.57 * cor_sigma(cor_sigma(cor_sigma(SSIM)))
  UM = 1.59 * cor_sigma(cor_sigma(MS_SSIM))
  UM = 1.10 * cor_sigma(cor_sigma(VIFP1))
  UM = 3.0 * (SMALLFRY * 0.01 - 0.8)
  UM = 1.46 * cor_sigma(SHARPENBAD)
  UM = 0.342 * (sqrt(sqrt(1.0 / NHW)) - 1.0)

    cor_sigma(M) = 1.0 - sqrt(1.0 - M * M)
```

#### Subsampling

The JPEG format allows for subsampling of the color channels to save space. For each 2x2 block of pixels per color channel (four pixels total) it can store four pixels (all of them), two pixels or a single pixel. By default, the JPEG encoder subsamples the non-luma channels to two pixels (often referred to as 4:2:0 subsampling). Most digital cameras do the same because of limitations in the human eye. This may lead to unintended behavior for specific use cases (see #12 for an example), so you can use `--subsample disable` to disable this subsampling.

#### Example Commands

```shell
# Default settings
jpeg-recompress image.jpg compressed.jpg

# High quality example settings
jpeg-recompress --quality high --min 60 image.jpg compressed.jpg

# Slow high quality settings (3-4x slower than above, slightly more accurate)
jpeg-recompress --accurate --quality high --min 60 image.jpg compressed.jpg

# Use SmallFry instead of SSIM
jpeg-recompress --method smallfry image.jpg compressed.jpg

# Use 4:4:4 sampling (disables subsampling).
jpeg-recompress --subsample disable image.jpg compressed.jpg

# Remove fisheye distortion (Tokina 10-17mm on APS-C @ 10mm)
jpeg-recompress --defish 2.6 --zoom 1.2 image.jpg defished.jpg

# Read from stdin and write to stdout with '-' as the filename
jpeg-recompress - - <image.jpg >compressed.jpg

# Convert RAW to JPEG via PPM from stdin
dcraw -w -q 3 -c IMG_1234.CR2 | jpeg-recompress --ppm - compressed.jpg

# Disable progressive mode (not recommended)
jpeg-recompress --no-progressive image.jpg compressed.jpg

# Disable all output except for errors
jpeg-recompress --quiet image.jpg compressed.jpg
```

### jpeg-compare

Compare two JPEG photos to judge how similar they are. The `fast` comparison method returns an integer from 0 to 99, where 0 is identical. PSNR, SSIM, and MS-SSIM return floats but require images to be the same dimensions.

```bash
# Do a fast compare of two images
jpeg-compare image1.jpg image2.jpg

# Calculate PSNR
jpeg-compare --method psnr image1.jpg image2.jpg

# Calculate SSIM
jpeg-compare --method ssim image1.jpg image2.jpg
```

### jpeg-hash

Create a hash of an image that can be used to compare it to other images quickly.

```bash
jpeg-hash image.jpg
```

### jpeg-zfpoint

Compress JPEG files by re-encoding them to the lowest JPEG quality using the peculiarity jpeg (zero point) quantization feature.

### webp-compress

Compress JPEGs by re-encoding to the smallest WEBP quality while keeping perceived visual quality the same.

This is a lossy operation, but the images are visually identical and it usually saves >50% of the size for JPEGs coming from a digital camera, particularly DSLRs.

All EXIF/IPTC/XMP and color profile metadata are not preserved!

Some basic photo-related editing options are available, such as removing fisheye lens distortion.

## Building

### Dependencies
 * [libiqa](https://github.com/ImageProcessing-ElectronicPublications/libiqa)
 * [libsmallfry](https://github.com/ImageProcessing-ElectronicPublications/libsmallfry)
 * [jpeg8](https://www.ijg.org)
 * [webp](https://developers.google.com/speed/webp/)

#### Debian

Debian users can install via `apt-get`:

```bash
sudo apt-get install build-essential autoconf pkg-config nasm libtool libjpeg8-dev
git clone https://github.com/ImageProcessing-ElectronicPublications/libiqa.git
cd libiqa
make
sudo make install
cd ..
git clone https://github.com/ImageProcessing-ElectronicPublications/libsmallfry.git
cd libsmallfry
make
sudo make install
cd ..
```

### Compiling (Linux and Mac OS X)

The `Makefile` should work as-is on Ubuntu and Mac OS X. Other platforms may need to set the location of `libjpeg.a` or make other tweaks.

```bash
make
```

### Installation

Install the binaries into `/usr/local/bin`:

```bash
sudo make install
```

## Links / Alternatives

* https://github.com/rflynn/imgmin
* https://news.ycombinator.com/item?id=803839

## License

* JPEG-Recompress is copyright &copy; 2015 Daniel G. Taylor
* Image Quality Assessment (IQA) is copyright 2011, Tom Distler (http://tdistler.com)
* SmallFry is copyright 2014, Derek Buitenhuis (https://github.com/dwbuiten)

All are released under an MIT license.

http://dgt.mit-license.org/


CHANGELOG
JPEG Recompress
https://github.com/ImageProcessing-ElectronicPublications/jpeg-recompress

2.6.4 "um"

    Research: correct UM model: all metrics.
    2023, zvezdochiot

2.6.3 "um"

    Research: correct UM model.
    2023, zvezdochiot

2.6.2 "sharpenbad"

    reimplementation SHARPENBAD:
      metric_sharpenbad

2.6.1 "msef"

    Added MSE and MSEF metric.
    Research: correct trends UM (44 images)

2.6.0 "sum"

    Added VIFP 1 layer in SUM metric.

2.5.9 "vifp1"

    Added VIFP 1 layer metric.

2.5.8 "neat"

    Neat adds up all values.

2.5.7 "bisection"

    Fix bisection

2.5.6 "subsample"

    Fix IPEP/#1: ignores subsampling option

2.5.5 "sigma"

    Move `sigma`-function from `libsmallfry` to `jmetrics.a`
    Research: correct trends UM (32 images)
    Remove metrics `cor` and `corsh`
    2022, zvezdochiot

2.5.4 "near"

    Research: correct trends UM (24 images)
    2022, zvezdochiot

2.5.3 "approx"

    Research: correct trends UM (16 images)
    2022, zvezdochiot

2.5.2 "UM"

    Research: correct trends UM (8 images)
    2022, zvezdochiot

2.5.1 "saved"

    Fix overflow saved.
    Correct quality scale (next = (1.0 + prev) / 2)
    Replacing multiplications with counters.
    2022, zvezdochiot

2.5.0 "webp"

    Use libwebp for WEBP format
    2022, zvezdochiot

2.4.0 "nhw"

    Update for libsmallfry 0.1.0
    metric_nhw
    2022, zvezdochiot

2.3.1 "weight"

    weight average for SUM metrics
    2021, zvezdochiot

2.3.0 "return"

    return from mozjpeg to jpeg8
    2021, zvezdochiot

2.2.1 "linearize"

    jpeg-recompress: linearize metric
    2018, zvezdochiot

2.2 "zfpoint"

    Find ZF point quality for JPEG
    Update for libsmallfry 0.0.3
    2018, zvezdochiot

2.1.3 "sum"

    jpeg-recompress: summary metric and linearize ssim
    2018, zvezdochiot

2.1.2

    Update for libsmallfry 0.0.2
    sharpenbad_metric
    2018, zvezdochiot

2.1.1

    fork jpeg-archive 2.1.1 to jpeg-recompress for compatible libsmallfry.
    Daniel G. Taylor & zvezdochiot


test/test.c
#include "../src/jmetrics.h"
#include "../src/test/describe.h"

describe ("Unit Tests", {
    it ("Should clamp values", {
        assert_equal_float(0.0, clamp(0.0, -10.0, 100.0));
        assert_equal_float(32.5, clamp(0.0, 32.5, 100.0));
        assert_equal_float(100.0, clamp(0.0, 150.3, 100.0));
    });

    it ("Should interpolate pixels", {
        unsigned char *image;
        int value;

        image = malloc(4);

        image[0] = 0;
        image[1] = 255;
        image[2] = 127;
        image[3] = 66;

        value = interpolate(image, 2, 1, 0.3, 0.6, 0);

        assert_equal(95, value);

        free(image);
    });

    it ("Should scale an image", {
        unsigned char *image;
        unsigned char *scaled;

        /*
        [  0  1  2  3
           4  5  6  7
           8  9 10 11
          12 13 14 15 ]
        */

        image = malloc(16);

        for (int x = 0; x < 16; x++) {
            image[x] = (unsigned char) x;
        }

        /*
        [  0  2
           8 10 ]
        */
        scale(image, 4, 4, &scaled, 2, 2);

        assert_equal(image[2], scaled[1]);
        assert_equal(image[8], scaled[2]);

        free(scaled);
        free(image);
    });

    it ("Should generate an image hash", {
        unsigned char *image;
        unsigned char *hash;

        image = malloc(16);

        /*
        [  0 15  2 13
           4 11  6  9
           8  7 10  5
          12  3 14  1 ]
        */

        for (int x = 0; x < 16; x++) {
            image[x] = (unsigned char) ((x % 2) ? 16 - x : x);
        }

        // Hash should be 101010100101010
        genHash(image, 4, 4, &hash);

        assert_equal(1, hash[0]);
        assert_equal(0, hash[1]);
        assert_equal(0, hash[5]);
        assert_equal(1, hash[9]);

        free(hash);
        free(image);
    });

    it ("Should calculate hamming distance", {
        int dist = hammingDist((unsigned char *) "101010", (unsigned char *) "111011", 6);
        assert_equal(2, dist);
    });

    it ("Should decode a PPM", {
        char *image = "P6\n2 2\n255\n\x1\x2\x3\x4\x5\x6\x7\x8\x9\xa\xb\xc";
        unsigned char *imageData;
        int width;
        int height;

        decodePpm((unsigned char *) image, 23, (unsigned char **) &imageData, &width, &height);

        assert_equal(2, width);
        assert_equal(2, height);
        assert_equal('\x1', imageData[0]);
        assert_equal('\xc', imageData[11]);

        free(imageData);
    })
});


src/jmetrics.h
/*
    Image editing functions
*/
#include <stdarg.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <sys/types.h>
#include <jpeglib.h>
#include <iqa.h>
#include <smallfry.h>

#ifdef _WIN32
#include <io.h>
#include <fcntl.h>
#endif

#ifndef JMETRICS_H
#define JMETRICS_H

#ifndef JMVERSION
#define JMVERSION "2.6.4"
#endif

#define MIN(a, b) ((a) < (b) ? (a) : (b))
#define MAX(a, b) ((a) > (b) ? (a) : (b))

// Subsampling method, which defines how much of the data from
// each color channel is included in the image per 2x2 block.
// A value of 4 means all four pixels are included, while 2
// means that only two of the four are included (hence the term
// subsampling). Subsampling works really well for photos, but
// can have issues with crisp colored borders (e.g. red text).
enum SUBSAMPLING_METHOD
{
    // Default is 4:2:0
    SUBSAMPLE_DEFAULT,
    // Using 4:4:4 is more detailed and will prevent fine text
    // from getting blurry (e.g. screenshots)
    SUBSAMPLE_444
};

enum filetype
{
    FILETYPE_UNKNOWN,
    FILETYPE_AUTO,
    FILETYPE_JPEG,
    FILETYPE_PPM
};

// Comparison method
enum METHOD
{
    UNKNOWN,
    FAST,
    MPE,
    MSE,
    PSNR,
    MSEF,
    SSIM,
    MS_SSIM,
    VIFP1,
    SMALLFRY,
    SHARPENBAD,
    COR,
    NHW,
    SSIMFRY,
    SSIMSHBAD,
    SUMMET
};

// Target quality (SSIM) value
enum QUALITY_PRESET
{
    LOW,
    MEDIUM,
    SUBHIGH,
    HIGH,
    VERYHIGH
};

/* Long command line options. */
enum longopts
{
    OPT_SHORT = 1000
};

/*
    Clamp a value between low and high.
*/
float clamp(float low, float value, float high);

/*
    Bilinear interpolation
*/
int interpolate(const unsigned char *image, int width, int components, float x, float y, int offset);

/*
    Get mean error per pixel rate.
*/
float metric_mpe(const unsigned char *original, const unsigned char *compressed, int width, int height, int components);
float metric_mse(const unsigned char *ref, const unsigned char *cmp, int width, int height, int channels);
float metric_stdev2(const unsigned char *ref, const unsigned char *cmp, int width, int height, int channels);
float metric_msef(const unsigned char *ref, const unsigned char *cmp, int width, int height, int channels);

/*
    Remove fisheye distortion from an image. The amount of distortion is
    controlled by strength, while zoom controls where the image gets
    cropped. For example, the Tokina 10-17mm ATX fisheye on a Canon APS-C
    body set to 10mm looks good with strength=2.6 and zoom=1.2.
*/
void defish(const unsigned char *input, unsigned char *output, int width, int height, int components, float strength, float zoom);

/*
    Convert an RGB image to grayscale. Assumes 8-bit color components,
    3 color components and a row stride of width * 3.
*/
unsigned long int grayscale(const unsigned char *input, unsigned char **output, int width, int height);

/*
    Generate an image hash given a filename. This is a convenience
    function which reads the file, decodes it to grayscale,
    scales the image, and generates the hash.
*/
int jpegHash(const char *filename, unsigned char **hash, int size);
int jpegHashFromBuffer(unsigned char *imageBuf, long bufSize, unsigned char **hash, int size);

/*
    Downscale an image with nearest-neighbor interpolation.
    http://jsperf.com/pixel-interpolation/2
*/
void scale(unsigned char *image, int width, int height, unsigned char **newImage, int newWidth, int newHeight);

/*
    Generate an image hash based on gradients.
    http://www.hackerfactor.com/blog/index.php?/archives/529-Kind-of-Like-That.html
*/
void genHash(unsigned char *image, int width, int height, unsigned char **hash);

/*
    Calculate the hamming distance between two hashes.
    http://en.wikipedia.org/wiki/Hamming_distance
*/
unsigned int hammingDist(const unsigned char *hash1, const unsigned char *hash2, int hashLength);

/* Print program version to stdout. */
void version(void);

/* Print an error message. */
void error(const char *format, ...);

/*
    Read a file into a buffer and return the length.
*/
unsigned long int readFile(char *name, void **buffer);

/*
    Decode a buffer into a JPEG image with the given pixel format.
    Returns the size of the image pixel array.
    See libjpeg.txt for a (very long) explanation.
*/
int checkJpegMagic(const unsigned char *buf, unsigned long int size);
unsigned long int decodeJpeg(unsigned char *buf, unsigned long int bufSize, unsigned char **image, int *width, int *height, int *jpegcs, int pixelFormat);

/*
    Decode buffer into a PPM image.
    Returns the size of the image pixel array.
*/
int checkPpmMagic(const unsigned char *buf, unsigned long int size);
unsigned long int decodePpm(unsigned char *buf, unsigned long int bufSize, unsigned char **image, int *width, int *height);

/*
    Encode a buffer of image pixels into a JPEG.
*/
unsigned long int encodeJpeg(unsigned char **jpeg, unsigned char *buf, int width, int height, int pixelFormat, int quality, int jpegcs, int progressive, int optimize, int subsample);

/* Automatically detect the file type of a given file. */
enum filetype detectFiletype(const char *filename);
enum filetype detectFiletypeFromBuffer(unsigned char *buf, unsigned long int bufSize);

/* Decode an image file with a given format. */
unsigned long int decodeFile(const char *filename, unsigned char **image, enum filetype type, int *width, int *height, int pixelFormat);
unsigned long int decodeFileFromBuffer(unsigned char *buf, unsigned long int bufSize, unsigned char **image, enum filetype type, int *width, int *height, int *jpegcs, int pixelFormat);

/*
    Get JPEG metadata (EXIF, IPTC, XMP, etc) and return a buffer
    with just this data, suitable for writing out to a new file.
    Reads in all APP0-APP15 markers as well as COM markers.
    Reads up to 20 markers.

    If comment is not NULL, then returns 1 if the comment is
    encountered, allowing scripts to detect if they have previously
    modified the file.
*/
int getMetadata(const unsigned char *buf, unsigned int bufSize, unsigned char **meta, unsigned int *metaSize, const char *comment);
FILE *openOutput(char *name);
void info(int quiet, const char *format, ...);

enum filetype parseInputFiletype(const char *s);
int parseSubsampling(const char *s);
enum QUALITY_PRESET parseQuality(const char *s);
float setTargetFromPreset(int preset);
enum METHOD parseMethod(const char *s);
float MetricRescale(int currentmethod, float value);
char* MetricName(int currentmethod);
float MetricCalc(int method, unsigned char *image1, unsigned char *image2, int width, int height, int components);
float MetricSigma(float cor);
int compareFastFromBuffer(unsigned char *imageBuf1, long bufSize1, unsigned char *imageBuf2, long bufSize2, int printPrefix, int size);
int compareFromBuffer(int method, unsigned char *imageBuf1, long bufSize1, unsigned char *imageBuf2, long bufSize2, int printPrefix, int umscale, enum filetype inputFiletype1, enum filetype inputFiletype2);
float waverage(float *x, int count);

#endif


src/jpeg-recompress.c
/*
    Recompress a JPEG file while attempting to keep visual quality the same
    by using structural similarity (SSIM) as a metric. Does a binary search
    between JPEG quality 40 and 95 to find the best match. Also makes sure
    that huffman tables are optimized if they weren't already.
*/

#include <getopt.h>
#include "jmetrics.h"

const char *COMMENT = "Compressed by jpeg-recompress";

void usage(char *progname)
{
    printf("usage: %s [options] input.jpg output.jpg\n\n", progname);
    printf("options:\n\n");
    printf("  -a, --accurate               favor accuracy over speed\n");
    printf("  -c, --no-copy                disable copying files that will not be compressed\n");
    printf("  -d, --defish [arg]           set defish strength [0.0]\n");
    printf("  -f, --force                  force process\n");
    printf("  -h, --help                   output program help\n");
    printf("  -l, --loops [arg]            set the number of runs to attempt [6]\n");
    printf("  -m, --method [arg]           set comparison method to one of:\n");
    printf("                               'mpe', 'psnr', 'mse', 'msef', 'cor', 'ssim', 'ms-ssim', 'vifp1',\n");
    printf("                               'smallfry', 'shbad', 'nhw', 'ssimfry', 'ssimshb', 'sum' [sum]\n");
    printf("  -n, --min [arg]              minimum JPEG quality [40]\n");
    printf("  -p, --no-progressive         disable progressive encoding\n");
    printf("  -q, --quality [arg]          set a quality preset: low, medium, subhigh, high, veryhigh [medium]\n");
    printf("  -r, --ppm                    parse input as PPM\n");
    printf("  -s, --strip                  strip metadata\n");
    printf("  -t, --target [arg]           set target quality [0.75]\n");
    printf("  -x, --max [arg]              maximum JPEG quality [98]\n");
    printf("  -z, --zoom [arg]             set defish zoom [1.0]\n");
    printf("  -Q, --quiet                  only print out errors\n");
    printf("  -S, --subsample [arg]        set subsampling method to one of 'default', 'disable' [default]\n");
    printf("  -T, --input-filetype [arg]   set input file type to one of 'auto', 'jpeg', 'ppm' [auto]\n");
    printf("  -V, --version                output program version\n");
    printf("  -Y, --ycbcr [arg]            YCbCr jpeg colorspace: 0 - source, >0 - YCrCb, <0 - RGB\n");
}

int main (int argc, char **argv)
{
    int method = SUMMET;

    // Number of binary search steps
    int attempts = 8;

    float target = 0.0f;
    int preset = MEDIUM;

    // Min/max JPEG quality
    int jpegMin = 1;
    int jpegMax = 99;

    int force = 0;
    int ycbcr = 0;
    // Strip metadata from the file?
    int strip = 0;

    // Disable progressive mode?
    int noProgressive = 0;

    // Defish the image?
    float defishStrength = 0.0f;
    float defishZoom = 1.0f;

    // Input format
    enum filetype inputFiletype = FILETYPE_AUTO;

    // Whether to copy files that cannot be compressed
    int copyFiles = 1;

    // Whether to favor accuracy over speed
    int accurate = 0;

    // Chroma subsampling method
    int subsample = SUBSAMPLE_DEFAULT;

    // Quiet mode (less output)
    int quiet = 0;

    unsigned char *buf, *original, *originalGray = NULL, *tmpImage;
    unsigned char *compressed = NULL, *compressedGray, *metaBuf;
    long bufSize = 0, originalSize = 0, originalGraySize = 0;
    long compressedGraySize = 0;
    unsigned long compressedSize = 0, saved = 0;
    int width, height, min, max, attempt, jpegcs, jpegcst;
    int quality, progressive, optimize, percent, app0_len;
    unsigned int metaSize = 0;
    float metric, umetric;
    char *inputPath, *outputPath;
    FILE *file;

    const char *optstring = "acd:fhl:m:n:pq:rst:x:z:QS:T:VY:";
    static const struct option opts[] =
    {
        { "accurate", no_argument, 0, 'a' },
        { "defish", required_argument, 0, 'd' },
        { "force", no_argument, 0, 'f' },
        { "help", no_argument, 0, 'h' },
        { "input-filetype", required_argument, 0, 'T' },
        { "loops", required_argument, 0, 'l' },
        { "max", required_argument, 0, 'x' },
        { "method", required_argument, 0, 'm' },
        { "min", required_argument, 0, 'n' },
        { "no-copy", no_argument, 0, 'c' },
        { "no-progressive", no_argument, 0, 'p' },
        { "ppm", no_argument, 0, 'r' },
        { "quality", required_argument, 0, 'q' },
        { "quiet", no_argument, 0, 'Q' },
        { "target", required_argument, 0, 't' },
        { "strip", no_argument, 0, 's' },
        { "subsample", required_argument, 0, 'S' },
        { "version", no_argument, 0, 'V' },
        { "ycbcr", required_argument, 0, 'Y' },
        { "zoom", required_argument, 0, 'z' },
        { 0, 0, 0, 0 }
    };
    int opt, longind = 0;

    char *progname = "jpeg-recompress";

    while ((opt = getopt_long(argc, argv, optstring, opts, &longind)) != -1)
    {
        switch (opt)
        {
        case 'a':
            accurate = 1;
            break;
        case 'c':
            copyFiles = 0;
            break;
        case 'd':
            defishStrength = atof(optarg);
            break;
        case 'f':
            force = 1;
            break;
        case 'h':
            usage(progname);
            return 0;
        case 'l':
            attempts = atoi(optarg);
            break;
        case 'm':
            method = parseMethod(optarg);
            break;
        case 'n':
            jpegMin = atoi(optarg);
            break;
        case 'p':
            noProgressive = 1;
            break;
        case 'q':
            preset = parseQuality(optarg);
            break;
        case 'r':
            inputFiletype = FILETYPE_PPM;
            break;
        case 's':
            strip = 1;
            break;
        case 't':
            target = atof(optarg);
            break;
        case 'x':
            jpegMax = atoi(optarg);
            break;
        case 'z':
            defishZoom = atof(optarg);
            break;
        case 'S':
            subsample = parseSubsampling(optarg);
            break;
        case 'T':
            if (inputFiletype != FILETYPE_AUTO)
            {
                error("multiple file types specified for the input file");
                return 1;
            }
            inputFiletype = parseInputFiletype(optarg);
            break;
        case 'Q':
            quiet = 1;
            break;
        case 'V':
            version();
            return 0;
        case 'Y':
            ycbcr = atoi(optarg);
            break;
        };
    }

    if (argc - optind != 2)
    {
        usage(progname);
        return 255;
    }

    inputPath = argv[optind];
    outputPath = argv[optind + 1];

    if (method == UNKNOWN)
    {
        error("invalid method!");
        usage(progname);
        return 255;
    }

    // No target passed, use preset!
    if (target < 0.001f)
    {
        target = setTargetFromPreset(preset);
    }

    /* Read the input into a buffer. */
    bufSize = readFile(inputPath, (void **) &buf);

    /* Detect input file type. */
    if (inputFiletype == FILETYPE_AUTO)
        inputFiletype = detectFiletypeFromBuffer(buf, bufSize);

    /*
     * Read original image and decode. We need the raw buffer contents and its
     * size to obtain meta data and the original file size later.
     */
    originalSize = decodeFileFromBuffer(buf, bufSize, &original, inputFiletype, &width, &height, &jpegcs, JCS_RGB);
    if (!originalSize)
    {
        error("invalid input file: %s", inputPath);
        return 1;
    }

    if (defishStrength)
    {
        info(quiet, "Defishing...\n");
        tmpImage = malloc(width * height * 3);
        defish(original, tmpImage, width, height, 3, defishStrength, defishZoom);
        free(original);
        original = tmpImage;
    }

    // Convert RGB input into Y
    originalGraySize = grayscale(original, &originalGray, width, height);

    if (inputFiletype == FILETYPE_JPEG)
    {
        // Read metadata (EXIF / IPTC / XMP tags)
        if (getMetadata(buf, bufSize, &metaBuf, &metaSize, COMMENT) && !force)
        {
            if (copyFiles)
            {
                info(quiet, "File already processed by jpeg-recompress!\n");
                file = openOutput(outputPath);
                if (file == NULL)
                {
                    error("could not open output file: %s", outputPath);
                    return 1;
                }

                fwrite(buf, bufSize, 1, file);
                fclose(file);

                free(buf);

                return 0;
            }
            else
            {
                error("file already processed by jpeg-recompress!");
                free(buf);
                return 2;
            }
        }
    }

    if (strip)
        metaSize = 0;
    else
        info(quiet, "Metadata size is %ukb\n", metaSize / 1024);

    if (!originalSize || !originalGraySize)
        return 1;

    if (ycbcr < 0)
        jpegcs = JCS_RGB;
    if (ycbcr > 0)
        jpegcs = JCS_YCbCr;

    if (jpegMin > jpegMax)
    {
        error("maximum JPEG quality must not be smaller than minimum JPEG quality!");
        return 1;
    }

    // Do a binary search to find the optimal encoding quality for the
    // given target SSIM value.
    min = jpegMin;
    max = jpegMax;
    for (attempt = attempts - 1; attempt >= 0; --attempt)
    {
        quality = (max + min + 1) / 2;

        /* Terminate early once bisection interval is a singleton. */
        if (min == max)
            attempt = 0;

        progressive = attempt ? 0 : !noProgressive;
        optimize = accurate ? 1 : (attempt ? 0 : 1);

        // Recompress to a new quality level, without optimizations (for speed)
        compressedSize = encodeJpeg(&compressed, original, width, height, JCS_RGB, quality, jpegcs, progressive, optimize, subsample);

        // Load compressed luma for quality comparison
        compressedGraySize = decodeJpeg(compressed, compressedSize, &compressedGray, &width, &height, &jpegcst, JCS_GRAYSCALE);

        if (!compressedGraySize)
        {
            error("unable to decode file that was just encoded!");
            return 1;
        }

        if (!attempt)
            info(quiet, "Final optimized ");

        // Measure quality difference
        metric = MetricCalc(method, originalGray, compressedGray, width, height, 1);
        umetric = MetricRescale(method, metric);
        info(quiet, MetricName(method));

        if (attempt)
            info(quiet, " at q=%i (%i - %i): UM %f\n", quality, min, max, umetric);
        else
            info(quiet, " at q=%i: UM %f\n", quality, umetric);

        if (umetric < target)
        {
            if (compressedSize >= bufSize)
            {
                free(compressed);
                free(compressedGray);

                if (copyFiles)
                {
                    info(quiet, "Output file would be larger than input!\n");
                    file = openOutput(outputPath);
                    if (file == NULL)
                    {
                        error("could not open output file: %s", outputPath);
                        return 1;
                    }

                    fwrite(buf, bufSize, 1, file);
                    fclose(file);

                    free(buf);

                    return 0;
                }
                else
                {
                    error("output file would be larger than input!");
                    free(buf);
                    return 1;
                }
            }
            min = MIN(quality, max);
        }
        else
        {
            max = MAX(quality, min);
        }

        // If we aren't done yet, then free the image data
        if (attempt)
        {
            free(compressed);
            free(compressedGray);
        }
    }

    free(buf);

    // Calculate and show savings, if any
    percent = (compressedSize + metaSize) * 100 / bufSize;
    saved = (bufSize > (compressedSize + metaSize)) ? (bufSize - compressedSize - metaSize) : 0;
    info(quiet, "New size is %i%% of original (saved %lu kb)\n", percent, saved / 1024);

    if (compressedSize >= bufSize && !force)
    {
        error("output file is larger than input, aborting!");
        return 1;
    }

    // Open output file for writing
    file = openOutput(outputPath);
    if (file == NULL)
    {
        error("could not open output file");
        return 1;
    }

    /* Check that the metadata starts with a SOI marker. */
    if (!checkJpegMagic(compressed, compressedSize))
    {
        error("missing SOI marker, aborting!");
        return 1;
    }

    /* Make sure APP0 is recorded immediately after the SOI marker. */
    if (compressed[2] != 0xff || (compressed[3] != 0xe0 && compressed[3] != 0xee))
    {
        error("missing APP0 marker, aborting!");
        return 1;
    }

    /* Write SOI marker and APP0 metadata to the output file. */
    app0_len = (compressed[4] << 8) + compressed[5];
    fwrite(compressed, 4 + app0_len, 1, file);

    /*
     * Write comment (COM metadata) so we know not to reprocess this file in
     * the future if it gets passed in again.
     */
    fputc(0xff, file);
    fputc(0xfe, file);
    fputc(0x00, file);
    fputc(strlen(COMMENT) + 2, file);
    fwrite(COMMENT, strlen(COMMENT), 1, file);

    /* Write additional metadata markers. */
    if (inputFiletype == FILETYPE_JPEG && !strip)
        fwrite(metaBuf, metaSize, 1, file);

    /* Write image data. */
    fwrite(compressed + 4 + app0_len, compressedSize - 4 - app0_len, 1, file);
    fclose(file);

    if (inputFiletype == FILETYPE_JPEG && !strip)
        free(metaBuf);

    free(compressed);
    free(original);
    free(originalGray);

    return 0;
}
